crate_universe/src/api/lockfile.rs

```diff
@@ -0,0 +1,220 @@
+#[serde(transparent)]
+struct CargoBazelLockfileImpl(Context);
+
+impl CargoBazelLockfile for CargoBazelLockfileImpl {
+    fn workspace_members(&self) -> BTreeSet<CrateId> {
+        self.0.workspace_members.keys().cloned().collect()
+    }
+
+    fn crate_info(&self, crate_id: &CrateId) -> Option<CrateInfo> {
+        let crate_context = self.0.crates.get(crate_id)?;
+        Some(CrateInfo {
+            name: crate_context.name.clone(),
+            version: crate_context.version.clone(),
+            library_target_name: crate_context.library_target_name.clone(),
+            is_proc_macro: crate_context
+                .targets
+                .iter()
+                .any(|t| matches!(t, Rule::ProcMacro(_))),
+            common_attributes: crate_context.common_attrs.clone(),
+        })
+    }
+}
+
+/// Information about a crate (which may be in-workspace or a dependency).
+#[derive(Deserialize, PartialEq, Eq, Debug)]
+pub struct CrateInfo {
+    name: String,
+    version: semver::Version,
+    library_target_name: Option<String>,
+    is_proc_macro: bool,
+
+    common_attributes: CommonAttributes,
+}
+
+impl CrateInfo {
+    /// The name of the crate.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// The version of the crate.
+    pub fn version(&self) -> &semver::Version {
+        &self.version
+    }
+
+    /// The name of the crate's root library target. This is the target that a dependent
+    /// would get if they were to depend on this crate.
+    pub fn library_target_name(&self) -> Option<&str> {
+        self.library_target_name.as_deref()
+    }
+
+    /// Whether the crate is a procedural macro.
+    pub fn is_proc_macro(&self) -> bool {
+        self.is_proc_macro
+    }
+
+    /// Dependencies required to compile the crate, without procedural macro dependencies.
+    pub fn normal_deps(&self) -> Select<BTreeSet<CrateDependency>> {
+        self.common_attributes.deps.clone()
+    }
+
+    /// Dependencies required to compile the tests for the crate, but not needed to compile the crate itself, without procedural macro dependencies.
+    pub fn dev_deps(&self) -> Select<BTreeSet<CrateDependency>> {
+        self.common_attributes.deps_dev.clone()
+    }
+
+    /// Procedural macro dependencies required to compile the crate.
+    pub fn proc_macro_deps(&self) -> Select<BTreeSet<CrateDependency>> {
+        self.common_attributes.proc_macro_deps.clone()
+    }
+
+    /// Procedural macro dependencies required to compile the tests for the crate, but not needed to compile the crate itself.
+    pub fn proc_macro_dev_deps(&self) -> Select<BTreeSet<CrateDependency>> {
+        self.common_attributes.proc_macro_deps_dev.clone()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{parse, CargoBazelLockfile};
+    use crate::config::CrateId;
+    use crate::context::crate_context::CrateDependency;
+    use semver::Version;
+    use std::collections::BTreeSet;
+
+    #[test]
+    fn test() {
```

Comment:
@illicitonion What is this testing?

Reply:
See https://github.com/bazelbuild/rules_rust/pull/2637 :)